main.c

```diff
@@ -462,22 +482,183 @@ static struct kanshi_config *read_config(void) {
 	return parse_config(config_path);
 }
 
+static void destroy_config(struct kanshi_config *config) {
+	struct kanshi_profile *profile, *tmp_profile;
+	wl_list_for_each_safe(profile, tmp_profile, &config->profiles, link) {
+		struct kanshi_profile_output *output, *tmp_output;
+		wl_list_for_each_safe(output, tmp_output, &profile->outputs, link) {
+			free(output->name);
+			wl_list_remove(&output->link);
+			free(output);
+		}
+		struct kanshi_profile_command *command, *tmp_command;
+		wl_list_for_each_safe(command, tmp_command, &profile->commands, link) {
+			free(command->command);
+			wl_list_remove(&command->link);
+			free(command);
+		}
+		wl_list_remove(&profile->link);
+		free(profile);
+	}
+	free(config);
+}
+
+static bool reload_config(struct kanshi_state *state) {
+	fprintf(stderr, "reloading config\n");
+	struct kanshi_config *config = read_config();
+	if (config != NULL) {
+		destroy_config(state->config);
+		state->config = config;
+		state->pending_profile = NULL;
+		state->current_profile = NULL;
+		return try_apply_profiles(state);
+	}
+	return false;
+}
+
+static int do_poll(struct pollfd *fds, nfds_t nfds) {
+	int ret;
+	do {
+		ret = poll(fds, nfds, -1);
+	} while (ret == -1 && errno == EINTR);
+	return ret;
+}
+
+enum readfds_type {
+	FD_WAYLAND		= 0,
+#ifdef KANSHI_HAS_VARLINK
+	FD_VARLINK		= 1,
+#endif
+	FD_COUNT
+};
+
+static int loop(struct kanshi_state *state) {
+	struct pollfd readfds[FD_COUNT];
+	readfds[FD_WAYLAND].fd = wl_display_get_fd(state->display);
+	readfds[FD_WAYLAND].events = POLLIN;
+#ifdef KANSHI_HAS_VARLINK
+	readfds[FD_VARLINK].fd = varlink_service_get_fd(state->service);
+	readfds[FD_VARLINK].events = POLLIN;
+#endif
+
+	struct pollfd writefds[1];
+	writefds[0].fd = readfds[FD_WAYLAND].fd;
+	writefds[0].events = POLLOUT;
+
+	while (state->running) {
+	 while (wl_display_prepare_read(state->display) != 0) {
+			if (wl_display_dispatch_pending(state->display) == -1) {
+				return EXIT_FAILURE;
+			}
+		}
+
+		int ret;
+
+		while (true) {
+			ret = wl_display_flush(state->display);
+			if (ret != -1 || errno != EAGAIN) {
+				break;
+			}
+
+			if (do_poll(writefds, ARRAY_LENGTH(writefds)) == -1) {
+				goto read_error;
+			}
+		}
+
+		if (ret < 0 && errno != EPIPE) {
+			goto read_error;
+		}
+
+		if (do_poll(readfds, ARRAY_LENGTH(readfds)) == -1) {
+			goto read_error;
+		}
+
+		if (wl_display_read_events(state->display) == -1) {
+			return EXIT_FAILURE;
+		}
+
+#ifdef KANSHI_HAS_VARLINK
+		if (readfds[FD_VARLINK].revents & POLLIN) {
+			varlink_return_if_fail(
+					varlink_service_process_events(state->service));
+		}
+#endif
+
+		if (wl_display_dispatch_pending(state->display) == -1) {
+			return EXIT_FAILURE;
+		}
+	}
+
+	return EXIT_SUCCESS;
+
+read_error:
+	wl_display_cancel_read(state->display);
+	return EXIT_FAILURE;
+}
+
+#ifdef KANSHI_HAS_VARLINK
```

Comment:
Instead of having this big `#ifdef`, maybe we can move this into a separate file?

Reply:
Moved into ipc.c.